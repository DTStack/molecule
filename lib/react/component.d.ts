import { GlobalEvent } from 'mo/common/event';
export interface IComponent<S = any> {
    setState(values: S, callback?: (prevState: S, nextState: S) => void): void;
    render(nextState?: S): void;
    onUpdateState(callback: (prevState: S, nextState: S) => void): void;
    getState(): S;
    /**
     * Subscribe the component event
     * @param name
     * @param callback
     */
    onEvent(name: any, callback: any): void;
}
export declare abstract class Component<S = any> extends GlobalEvent implements IComponent<S> {
    protected abstract state: S;
    private _event;
    constructor();
    /**
     * Set the state values, and notify the view component to re render
     * @param values update target state values
     */
    setState(values: S, callback?: (prevState: S, nextState: S) => void): void;
    /**
     * Initiative notify the component to render the view by the state
     * @param nextState
     */
    render(nextState?: S): void;
    onUpdateState(callback: (prevState: S, nextState: S) => void): void;
    getState(): S;
    onEvent(name: any, callback: any): void;
}
