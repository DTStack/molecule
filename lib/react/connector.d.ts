import * as React from 'react';
import { IComponent } from './component';
import { Controller } from './controller';
export declare type ServiceObject = {
    [index: string]: IComponent;
};
export declare type ControllerObject = {
    [index: string]: Controller;
};
export declare function connect(Service: IComponent | ServiceObject, View: React.ComponentType<any>, Controller?: Controller | ControllerObject, watchFiled?: object): {
    new (props: any): {
        state: {
            lastUpdated: number;
        };
        componentDidMount(): void;
        onChange(prevState: any, nextState: any): void;
        update: () => void;
        getServiceState(): {};
        render(): JSX.Element;
        context: any;
        setState<K extends never>(state: {} | ((prevState: Readonly<{}>, props: Readonly<{}>) => {} | Pick<{}, K> | null) | Pick<{}, K> | null, callback?: (() => void) | undefined): void;
        forceUpdate(callback?: (() => void) | undefined): void;
        readonly props: Readonly<{}> & Readonly<{
            children?: React.ReactNode;
        }>;
        refs: {
            [key: string]: React.ReactInstance;
        };
        shouldComponentUpdate?(nextProps: Readonly<{}>, nextState: Readonly<{}>, nextContext: any): boolean;
        componentWillUnmount?(): void;
        componentDidCatch?(error: Error, errorInfo: React.ErrorInfo): void;
        getSnapshotBeforeUpdate?(prevProps: Readonly<{}>, prevState: Readonly<{}>): any;
        componentDidUpdate?(prevProps: Readonly<{}>, prevState: Readonly<{}>, snapshot?: any): void;
        componentWillMount?(): void;
        UNSAFE_componentWillMount?(): void;
        componentWillReceiveProps?(nextProps: Readonly<{}>, nextContext: any): void;
        UNSAFE_componentWillReceiveProps?(nextProps: Readonly<{}>, nextContext: any): void;
        componentWillUpdate?(nextProps: Readonly<{}>, nextState: Readonly<{}>, nextContext: any): void;
        UNSAFE_componentWillUpdate?(nextProps: Readonly<{}>, nextState: Readonly<{}>, nextContext: any): void;
    };
    contextType?: React.Context<any> | undefined;
};
