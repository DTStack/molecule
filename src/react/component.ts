import { cloneDeep } from 'lodash';
import { EventEmitter, GlobalEvent } from 'mo/common/event';

export enum ComponentEvents {
    Update = 'Component.Update',
}

export interface IComponent<S = any> {
    /**
     * Set the Component state
     * @param values The next values of state
     * @param callback calling after setState
     */
    setState(values: S, callback?: (prevState: S, nextState: S) => void): void;
    /**
     * Trigger the Component update event
     * @param nextState
     */
    render(nextState?: S): void;
    /**
     * Listen to the Component state update event
     * @param listener
     */
    onUpdateState(listener: (prevState: S, nextState: S) => void): void;
    /**
     * Remove the Component update event listening, default is remove all,
     * also you can remove one by pass the listener
     */
    removeOnUpdateState(listener?: Function): void;
    /**
     * Force to update the Component
     */
    forceUpdate(): void;
    /**
     * Get the Component state
     */
    getState(): S;
}

export abstract class Component<S = any>
    extends GlobalEvent
    implements IComponent<S>
{
    protected abstract state: S;
    private _event: EventEmitter;

    constructor() {
        super();
        this._event = new EventEmitter();
    }

    /**
     * Set the state values, and notify the view component to re render
     * @param values update target state values
     */
    public setState(
        values: Partial<S>,
        callback?: (prevState: S, nextState: S) => void
    ) {
        const nextState = Object.assign(this.state, values);
        this.render(nextState);
        callback?.(this.state, nextState);
    }

    /**
     * Initiative notify the component to render the view by the state
     * @param nextState
     */
    public render(nextState?: S) {
        this._event.emit(ComponentEvents.Update, this.state, nextState);
    }

    public onUpdateState(listener: (prevState: S, nextState: S) => void) {
        this._event.subscribe(ComponentEvents.Update, listener);
    }

    public removeOnUpdateState(listener?: Function): void {
        this._event.unsubscribe(ComponentEvents.Update, listener);
    }

    public forceUpdate() {
        this.setState(cloneDeep(this.state));
    }

    public getState(): S {
        return this.state;
    }
}
