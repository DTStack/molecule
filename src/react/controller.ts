import { GlobalEvent } from 'mo/common/event';

export abstract class Controller extends GlobalEvent {
    public abstract initView(): void;
}
