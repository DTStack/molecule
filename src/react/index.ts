export * from './component';
export * from './helper';
export * from './connector';
export * from './controller';
