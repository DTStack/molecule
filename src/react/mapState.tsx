import React, { Component } from 'react';
import Logger from 'mo/common/logger';

/**
 * Mapping the state to the component
 * TODO support mapping service method to the component.
 * @param WrappedComponent The component will be wrapped
 * @param state The state you want to injected, notice the state must be an observable object
 * @param subscribes The events of your subscribe, it used to trigger the state re render
 */
export function mapState<S, T>(
    WrappedComponent: React.ComponentType<S>,
    state: S,
    actions?: T
) {
    return class StateProvider extends Component {
        state: { lastUpdated: number };
        constructor(props) {
            super(props);
            this.onChange = this.onChange.bind(this);
            this.state = {
                lastUpdated: Date.now(),
            };
        }

        componentDidMount() {
            // There is no declare state parameter as IObservable type, so must convert to any type.
            if ((state as any).subscribe) {
                (state as any).subscribe(this.onChange);
            } else {
                Logger.error(
                    'The state parameter of mapState must be an observable object.'
                );
            }
        }
        /**
         * TODO: Performance optimize, now whatever any properties changed in target,
         * there always be trigger the onChange event, so need a compare operation.
         * @param nextState changed data
         */
        onChange(nextState: S) {
            Logger.info(nextState, state);
            this.setState({
                lastUpdated: Date.now(),
            });
        }

        render() {
            return (
                <WrappedComponent
                    {...this.state}
                    {...state}
                    {...this.props}
                    {...actions}
                />
            );
        }
    };
}
