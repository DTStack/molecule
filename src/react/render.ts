import * as ReactDOM from 'react-dom';
import { createRoot, Root } from 'react-dom/client';

const reactdom: typeof ReactDOM & {
    createRoot?: typeof createRoot;
    __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED?: {
        usingClientEntryPoint?: boolean;
    };
} = {
    ...ReactDOM,
};

let nextReactDOMClient: typeof createRoot | undefined;

export const renderedSign = Symbol('__molecule__root');

if (ReactDOM.version.startsWith('18')) {
    nextReactDOMClient = reactdom.createRoot;
}

function toggleWarning(skip: boolean) {
    const { __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED } = reactdom;

    if (
        __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED &&
        typeof __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED === 'object'
    ) {
        __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.usingClientEntryPoint =
            skip;
    }
}

export const render = (
    node: JSX.Element,
    container: HTMLElement & {
        [renderedSign]?: Root;
    }
) => {
    if (nextReactDOMClient) {
        toggleWarning(true);
        const root = container[renderedSign] || nextReactDOMClient(container);
        toggleWarning(false);
        root.render(node);

        container[renderedSign] = root;
        return;
    }

    reactdom.render(node, container);
};

export const unmout = (
    container: HTMLElement & {
        [renderedSign]?: Root;
    }
) => {
    if (nextReactDOMClient) {
        container[renderedSign]?.unmount();
        Reflect.deleteProperty(container, renderedSign);
        return true;
    }

    return reactdom.unmountComponentAtNode(container);
};
