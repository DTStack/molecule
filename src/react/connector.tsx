import 'reflect-metadata';
import { container } from 'tsyringe';
import React, { Component } from 'react';
import Logger from 'mo/common/logger';
import { IComponent } from './component';
import { Controller } from './controller';

export type ServiceObject = {
    [index: string]: IComponent;
};

export type ControllerObject = {
    [index: string]: Controller;
};

export function connect<T = any>(
    Service: IComponent | ServiceObject,
    View: React.ComponentType<any>,
    Controller?: Controller | ControllerObject
): React.ComponentType<T> {
    return class Connector extends Component<T, any> {
        state: { lastUpdated: number };
        private _isMounted = false;
        constructor(props) {
            super(props);
            this.onChange = this.onChange.bind(this);
            this.state = {
                lastUpdated: Date.now(),
            };
        }

        componentDidMount() {
            this._isMounted = true;
            this.handleService((service) => {
                service.onUpdateState(this.onChange);
            });
        }

        componentWillUnmount() {
            this._isMounted = false;
            this.handleService((service) => {
                service.removeOnUpdateState(this.onChange);
            });
        }

        onChange(prevState, nextState) {
            Logger.info(prevState, nextState, (container as any)._registry);
            this.update();
        }

        update = () => {
            if (this._isMounted) {
                this.setState({
                    lastUpdated: Date.now(),
                });
            }
        };

        getServiceState() {
            const target = {};
            this.handleService((service, prop) => {
                if (prop) {
                    Object.assign(target, {
                        [prop]: { ...service.getState() },
                    });
                } else {
                    Object.assign(target, { ...service.getState() });
                }
            });
            return target;
        }

        handleService(callback: (service: IComponent, prop?: string) => void) {
            if (this.isValidService(Service as IComponent)) {
                callback(Service as IComponent);
            } else {
                for (const name in Service) {
                    if (name) {
                        const service: IComponent = Service[name];
                        if (this.isValidService(service)) {
                            callback(service, name);
                        }
                    }
                }
            }
        }

        getControllerProperties() {
            const target = {};
            if (!Controller) {
                return target;
            }

            this.handleController((controller: Controller, prop) => {
                if (prop) {
                    Object.assign(target, {
                        [prop]: this.getObjectPublicProperties(controller),
                    });
                } else {
                    Object.assign(
                        target,
                        this.getObjectPublicProperties(controller)
                    );
                }
            });
            return target;
        }

        handleController(
            callback: (controller: Controller, prop?: string) => void
        ) {
            if (this.isValidController(Controller as Controller)) {
                callback(Controller as Controller);
            } else {
                for (const name in Controller) {
                    if (name) {
                        const controller: Controller = Controller[name];
                        if (this.isValidController(controller)) {
                            callback(controller, name);
                        }
                    }
                }
            }
        }

        render() {
            return (
                <View
                    {...this.state}
                    {...this.props}
                    {...this.getServiceState()}
                    {...this.getControllerProperties()}
                />
            );
        }

        private isValidService(service: IComponent) {
            return typeof service.onUpdateState === 'function';
        }

        private isValidController(controller: Controller) {
            return typeof controller.initView === 'function';
        }

        private getObjectPublicProperties(obj: Controller) {
            const keys = Object.keys(obj);
            const result = {};
            keys.forEach((key) => {
                // Filter the Service, Controller instances and private properties which start with '_'
                // TODO There need a better way to filter the private properties of Object,
                // maybe we can make use of the # identifier in future.
                if (
                    !key.endsWith('Service') &&
                    !key.endsWith('Controller') &&
                    !key.startsWith('_')
                ) {
                    result[key] = obj[key];
                }
            });
            return result;
        }
    };
}
