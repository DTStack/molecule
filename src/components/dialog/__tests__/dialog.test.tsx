import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';

import ConfirmDialog from '../confirmDialog';
import { Modal, ConfirmState } from '../index';

import {
    wrapDialogClassName,
    iconConfirmClassName,
    textConfirmClassName,
    detailConfirmClassName,
} from '../base';

const { confirm, warn, warning } = Modal;

const TEST_ID = 'test-id';

describe('Test Modal Component', () => {
    test('Match the Modal snapshot', () => {
        const component = renderer.create(<Modal />);
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('Modal component default value', () => {
        const CANCEL_FN = jest.fn();
        const OK_FN = jest.fn();
        const TEST_OK = 'ok';
        const TEST_CANCEL = 'cancel';
        const BUTTON_OK = 'button-ok';
        const BUTTON_CANCEL = 'button-cancel';
        const okButtonProps = {
            'data-testid': BUTTON_OK,
            className: BUTTON_CANCEL,
        };
        const cancelButtonProps = {
            'data-testid': BUTTON_CANCEL,
            className: BUTTON_OK,
        };
        const wrapper = render(
            <Modal
                visible={true}
                okText={TEST_OK}
                cancelText={TEST_CANCEL}
                okButtonProps={okButtonProps as any}
                cancelButtonProps={cancelButtonProps as any}
                onOk={OK_FN}
                onCancel={CANCEL_FN}
            />
        );
        const buttonOK = wrapper.getByTestId(BUTTON_OK);
        const buttonCancel = wrapper.getByTestId(BUTTON_CANCEL);

        expect(buttonOK.textContent).toBe(TEST_OK);
        expect(buttonCancel.textContent).toBe(TEST_CANCEL);
        expect(buttonOK.classList).toContain(BUTTON_CANCEL);
        expect(buttonCancel.classList).toContain(BUTTON_OK);

        fireEvent.click(buttonOK);
        expect(OK_FN).toBeCalled();

        fireEvent.click(buttonCancel);
        expect(CANCEL_FN).toBeCalled();
    });

    test('Modal component custom footer', () => {
        const TEST_FOOTER = <a data-testid={TEST_ID}>{TEST_ID}</a>;
        const wrapper = render(<Modal footer={TEST_FOOTER} visible={true} />);
        const footer = wrapper.getByTestId(TEST_ID);

        expect(footer.textContent).toBe(TEST_ID);
    });

    test('Modal component centered', () => {
        const TEST_FOOTER = <a data-testid={TEST_ID}>{TEST_ID}</a>;
        const wrapper = render(
            <Modal footer={TEST_FOOTER} centered={true} visible={true} />
        );
        const component = wrapper.getByTestId(TEST_ID);
        const div = component?.closest('div')!.parentElement!.parentElement
            ?.parentElement;

        expect(div!.classList).toContain(wrapDialogClassName);
    });
});

describe('Test Confirm Component', () => {
    function $$(className) {
        return document.body.querySelector(className);
    }

    test('Confirm general display', async () => {
        const className = iconConfirmClassName(ConfirmState.confirm);
        const DESTROY_ID = 'test-destroy';
        const TEST_TITLE = 'test-title';
        const TEST_CONTENT = 'test-content';
        const CANCEL_FN = jest.fn();
        const OK_FN = jest.fn();
        const TEST_CONFIRM = 'test-confirm';

        class App extends React.Component {
            state = {
                destroy: () => {},
            };
            openConfirm = () => {
                const { destroy } = confirm({
                    title: TEST_TITLE,
                    content: TEST_CONTENT,
                    'data-testid': TEST_CONFIRM,
                    onOk: OK_FN,
                    onCancel: CANCEL_FN,
                } as any);

                this.setState({
                    destroy,
                });
            };

            render() {
                const { destroy } = this.state;
                return (
                    <>
                        <div
                            data-testid={TEST_ID}
                            onClick={() => this.openConfirm()}
                        >
                            click
                        </div>
                        <div data-testid={DESTROY_ID} onClick={() => destroy()}>
                            destroy
                        </div>
                    </>
                );
            }
        }

        const wrapper = render(<App />);

        fireEvent.click(wrapper.getByTestId(TEST_ID));
        await waitFor(async () => {
            const component = $$(`div[class*="${className}"]`);
            expect(component!.classList).toContain(className);

            const icon = $$('.codicon-warning');
            expect(icon).not.toBeNull();

            const title = $$('.mo-modal-title');
            expect(title?.textContent).toBe(TEST_TITLE);

            const confirmTitle = $$(`.${textConfirmClassName}`);
            expect(confirmTitle?.textContent).toBe(TEST_TITLE);

            const confirmDetail = $$(`.${detailConfirmClassName}`);
            expect(confirmDetail?.textContent).toBe(TEST_CONTENT);

            /**
             * Destroy the confirm
             */
            fireEvent.click(wrapper.getByTestId(DESTROY_ID));
            await waitFor(() => {
                const component = $$(`div[class*="${className}"]`);
                expect(component).toBeNull();
            });
        });
    });

    test('Warn confirm general display', async () => {
        const className = iconConfirmClassName(ConfirmState.warning);

        class App extends React.Component {
            state = {
                distory: null,
            };
            openConfirm = () => {
                warn({});
            };

            render() {
                return (
                    <div
                        data-testid={TEST_ID}
                        onClick={() => this.openConfirm()}
                    >
                        click
                    </div>
                );
            }
        }

        const wrapper = render(<App />);

        fireEvent.click(wrapper.getByTestId(TEST_ID));
        await waitFor(() => {
            const confirmDetail = document.body.querySelector(`.${className}`);
            expect(confirmDetail?.textContent).not.toBeNull();
        });
    });

    test('Warning confirm general display', async () => {
        const className = iconConfirmClassName(ConfirmState.warning);

        class App extends React.Component {
            state = {
                distory: null,
            };
            openConfirm = () => {
                warning({});
            };

            render() {
                return (
                    <div
                        data-testid={TEST_ID}
                        onClick={() => this.openConfirm()}
                    >
                        click
                    </div>
                );
            }
        }

        const wrapper = render(<App />);

        fireEvent.click(wrapper.getByTestId(TEST_ID));
        await waitFor(() => {
            const confirmDetail = document.body.querySelector(`.${className}`);
            expect(confirmDetail?.textContent).not.toBeNull();
        });
    });
});

describe('Test confirmDialog', () => {
    test('Match ConfirmDialog Snapshot', () => {
        const component = renderer.create(
            <ConfirmDialog
                close={() => {}}
                title="Are you sure you want to permanently delete ?"
                content="This action is irreversible!"
            ></ConfirmDialog>
        );
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('The ConfirmDialog Snapshot', () => {
        const component = renderer.create(
            <ConfirmDialog
                close={() => {}}
                title="Are you sure you want to permanently delete ?"
                content="This action is irreversible!"
            ></ConfirmDialog>
        );
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('The ConfirmDialog close and destroy', async () => {
        const CLOSE_FN = jest.fn();
        const BUTTON_OK = 'button-ok';
        const BUTTON_CANCEL = 'button-cancel';
        const cancelButtonProps = {
            'data-testid': BUTTON_CANCEL,
            className: BUTTON_OK,
        };
        const wrapper = render(
            <ConfirmDialog
                cancelButtonProps={cancelButtonProps}
                onCancel={CLOSE_FN}
                close={CLOSE_FN}
                okCancel={true}
                visible={true}
                title="Are you sure you want to permanently delete ?"
                content="This action is irreversible!"
            ></ConfirmDialog>
        );
        const component = wrapper.getByTestId(BUTTON_CANCEL);

        fireEvent.click(component);
        expect(CLOSE_FN).toBeCalled();
    });

    test('The ConfirmDialog actions', async () => {
        const ACTIONS = <a data-testid={TEST_ID}>molecule</a>;
        const wrapper = render(
            <ConfirmDialog
                actions={ACTIONS}
                close={() => {}}
                okCancel={true}
                visible={true}
                title="Are you sure you want to permanently delete ?"
                content="This action is irreversible!"
            ></ConfirmDialog>
        );
        const component = wrapper.getByTestId(TEST_ID);

        expect(component).not.toBeNull();
    });
});
