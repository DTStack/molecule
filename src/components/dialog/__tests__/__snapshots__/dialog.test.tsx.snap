// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Modal Component Match the Modal snapshot 1`] = `null`;

exports[`Test confirmDialog Match ConfirmDialog Snapshot 1`] = `null`;

exports[`Test confirmDialog The ConfirmDialog Snapshot 1`] = `null`;
